File: src-engine/engine/managers/ShipManager.java
package engine.managers;

import java.util.List;

import javax.media.opengl.GL2;

import engine.math.Distance;
import engine.math.Radius;
import engine.math.Unit;
import model.Coords;
import model.ship.PlayerShip;
import model.ship.Ship;
import engine.calculators.CoordsCalculator;
import engine.calculators.DegreeCalculator;
import engine.calculators.VelocityCalculator;
import game.architecture.Constants;
import game.architecture.Manager;
import model.ship.parts.Control;
import model.ship.parts.Engine;
import model.ship.parts.Position;
import model.type.DrawableResolution;

/**
 * Manages the {@link Ship} objects. Updates positions and draws them.
 */
public class ShipManager extends Manager<Ship> {

    /* ========== PUBLIC ========== */
    /**
     * Draws the given {@link Ship} list using given {@link GL2} graphic context.
     * 
     * @param gl - {@link GL2} graphical context of drawing.
     * @param ships - {@link List} of {@link Ship}s to be drawn.
     */
    public void draw(GL2 gl, List<Ship> ships) {
        for (Ship ship : ships) {
            shipDrawer.draw(gl, ship);
        }
    }

    /**
     * Executes {@link Ship}s tasks, checks collisions and resolution and updates position.
     * 
     * @param ships - {@link List} of {@link Ship}s to process.
     * @param playerShip - used to calculate resolution and collisions.
     */
    public void tick(List<Ship> ships, PlayerShip playerShip) {
        updateCalculations(playerShip);

        for (Ship ship : ships) {
            updateDistance(ship, playerShip);
            updateCalculations(ship);
        }
    }

    /* ========== PRIVATE ========== */
    private void updateCalculations(Ship ship) {

        // Parameters
        Position position = ship.getPosition();
        Control control = ship.getControl();
        Engine engine = ship.getEngine();

        // Calculations
        DegreeCalculator.rotateShip(position, engine);
        VelocityCalculator.setVelocity(engine, control);
        CoordsCalculator.setShipPosition(position, engine);
        VelocityCalculator.decreaseAcceleration(engine);
    }

    private void updateDistance(Ship ship, PlayerShip playerShip) {

        // Parameters
        Coords playerShipCoords = playerShip.getPosition().getCoords();
        Coords shipCoords = ship.getPosition().getCoords();
        double size = ship.getSize();

        // Distance
        double distance = shipCoords.distanceTo(playerShipCoords);
        setResolution(ship, size, distance);

        // TODO collisions
    }

    private void setResolution(Ship ship, double size, double distance) {
        Radius radius = new Radius((float) (size * Constants.SHIP_SIZE_MAGNIFIER), Unit.KM);
        Distance distance1 = new Distance((float) distance, Unit.AU);
        DrawableResolution resolution = DrawableResolution.Companion.determineResolution(distance1.div(radius));
        ship.setResolution(resolution);
    }
}

File: src-engine/engine/managers/ShipManager.kt
package engine.managers

import engine.calculators.VelocityCalculator
import engine.math.Distance
import engine.math.Radius
import engine.math.ScaleUnit
import game.architecture.Constants
import game.architecture.GameComponentContainer
import game.architecture.Manager
import model.ship.PlayerShip
import model.ship.Ship
import model.ship.parts.Engine
import model.ship.parts.Position
import model.type.DrawableResolution
import java.lang.Math.cos
import java.lang.Math.sin
import javax.media.opengl.GL2

/**
 * Manages the [Ship] objects. Updates positions and draws them.
 */
class ShipManager : Manager<Ship>() {

    /* ========== PUBLIC ========== */
    /**
     * Draws the given [Ship] list using given [GL2] graphic context.
     *
     * @param gl - [GL2] graphical context of drawing.
     * @param ships - [List] of [Ship]s to be drawn.
     */
    fun draw(gl: GL2, ships: List<Ship>) {
        for (ship in ships) {
            GameComponentContainer.shipDrawer.draw(gl, ship)
        }
    }

    /**
     * Executes [Ship]s tasks, checks collisions and resolution and updates position.
     *
     * @param ships - [List] of [Ship]s to process.
     * @param playerShip - used to calculate resolution and collisions.
     */
    fun tick(ships: List<Ship>, playerShip: PlayerShip) {
        updateCalculations(playerShip)

        for (ship in ships) {
            updateDistance(ship, playerShip)
            updateCalculations(ship)
        }
    }

    /* ========== PRIVATE ========== */
    private fun updateCalculations(ship: Ship) {

        // Parameters
        val position = ship.position
        val control = ship.control
        val engine = ship.engine

        // Calculations
        rotateShip(position, engine)
        VelocityCalculator.setVelocity(engine, control)
        setShipPosition(position, engine)
        VelocityCalculator.decreaseAcceleration(engine)
    }

    fun rotateShip(position: Position, engine: Engine) {
        position.rotationX = position.rotationX.plus(engine.rotationXChange)
        position.rotationY = position.rotationY.plus(engine.rotationYChange)
    }

    private fun updateDistance(ship: Ship, playerShip: PlayerShip) {

        // Parameters
        val playerShipCoords = playerShip.position.coords
        val shipCoords = ship.position.coords

        // Distance
        val distance = shipCoords.distanceTo(playerShipCoords).toDouble()
        setResolution(ship, distance)

        // TODO collisions
    }

    private fun setResolution(ship: Ship, distance: Double) {
        val radius = Radius((ship.size * Constants.SHIP_SIZE_MAGNIFIER).toFloat(), ScaleUnit.KM)
        val distance1 = Distance(distance.toFloat(), engine.math.ScaleUnit.AU)
        val resolution = DrawableResolution.determineResolution(distance1.div(radius))
        ship.resolution = resolution
    }

    private fun setShipPosition(position: Position, engine: Engine) {
        val velocity = engine.velocity
        val coords = position.coords

        val rotationX = position.rotationX.toRadians()
        val rotationY = position.rotationY.toRadians().negate()

        coords.x = coords.x + velocity * Math.cos(rotationX.value().toDouble()).toFloat()
        coords.y = coords.y + velocity * Math.sin(rotationY.value().toDouble()).toFloat()
        coords.z = coords.z + velocity * Math.sin(rotationX.value().toDouble()).toFloat()
    }
}